src/components/createContainer.jsx

```diff
@@ -32,41 +34,47 @@ export default function createContainer(DecoratedComponent, specs) {
     static DecoratedComponent = DecoratedComponent
 
     static contextTypes = {
-      renderLoading: PropTypes.func.isRequired,
-      adaptor: adaptorShape.isRequired,
+      adrenaline: PropTypes.shape({
+          renderLoading: PropTypes.func.isRequired,
+          adaptor: adaptorShape.isRequired
+      }).isRequired
+    }
+
+    static childContextTypes = {
+        adrenaline: containerShape.isRequired
     }
 
     constructor(props, context) {
       super(props, context);
-
       this.isNotMounted = true;
+      this.state = {};
+    }
 
-      const { adaptor } = context;
-      //DecoratedComponent.prototype.shouldComponentUpdate = adaptor.shouldComponentUpdate;
-
-      const initialArgs = !!specs.args ? specs.args(props) : {};
-
-      this.state = {
-        data: {},
-        args: initialArgs,
-      };
-
-      const dataKeys = Object.keys(specs.queries(initialArgs));
-      const { propTypes } = DecoratedComponent;
-      const requiredKeys = dataKeys.filter(key =>
-        propTypes[key] && !propTypes[key].isRequired
-      );
-
-      this.isDataLoaded = (data = {}) => requiredKeys.every(key => {
-        return data[key] !== null && data[key] !== undefined
-      });
+    getChildContext() {
+        const { adrenaline } = this.context;
+        const container = {
+            setArgs: this.setArgs,
+            state: this.state
+        };
+        return {
+            adrenaline: {
+                ...adrenaline,
+                container
+            }
+        };
     }
 
     componentDidMount() {
-      const { adaptor } = this.context;
+      const adaptor = this.getAdaptor();
       this.isNotMounted= false;
+      this.setArgs(this.computeArgs(this.props));
```

Comment:
I'm sorry, I don't understand. Are you suggesting we get rid of the ```setArgs``` passed to decorated components? If so, I think yes. ```setArgs``` makes sense if you support 'initialArgs' (like before) and components needs to react to changes based on their state (as opposed to solely their props). This might be a valid use case, but I suppose the idomatic redux way would be to keep that state in the store. In our application, we've wrapped the Adrenaline container in a [react-redux connect](https://github.com/rackt/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options)'ed component so the state flows in to Adrenaline as props.

It might make sense to keep 'initialArgs' and 'setArgs' around, but I would force 'setArgs' to be accessible via container context ```this.context.adrenaline.container.setArgs``` to discourage it. Adrenaline should avoid from passing interface props to decorated components.